core_impl/java/com/google/instrumentation/stats/StatsSerializer.java

```diff
@@ -14,77 +14,71 @@
 package com.google.instrumentation.stats;
 
 import com.google.instrumentation.stats.proto.StatsContextProto;
+import com.google.io.base.VarInt;
+import com.google.protobuf.ByteString;
 
 import java.io.InputStream;
 import java.io.IOException;
 import java.io.OutputStream;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
 import java.util.HashMap;
 import java.util.Map.Entry;
+import java.util.Set;
 
 /**
  * Native implementation {@link StatsContext} serialization.
  */
 final class StatsSerializer {
-  private static final char TAG_PREFIX = '\2';
-  private static final char TAG_DELIM = '\3';
-
-
   // Serializes a StatsContext by transforming it into a StatsContextProto. The
-  // encoded tags are of the form: (<tag prefix> + 'key' + <tag delim> + 'value')*
+  // encoded tags are of the form:
+  //   num_tags [key_len key_bytes value_len value_bytes]*
   static void serialize(StatsContextImpl context, OutputStream output) throws IOException {
-    StringBuilder builder = new StringBuilder();
-    for (Entry<String, String> tag : context.tags.entrySet()) {
-      builder
-          .append(TAG_PREFIX)
-          .append(tag.getKey())
-          .append(TAG_DELIM)
-          .append(tag.getValue());
+    ByteBuffer buffer = ByteBuffer.allocate(1024);
+    Set<Entry<String, String>> tags = context.tags.entrySet();
+    VarInt.putVarInt(tags.size(), buffer);
+    for (Entry<String, String> tag : tags) {
+      encode(tag.getKey(), buffer);
+      encode(tag.getValue(), buffer);
     }
-    StatsContextProto.StatsContext.newBuilder().setTags(builder.toString()).build().writeTo(output);
+    int length = buffer.capacity() - buffer.remaining();
+    buffer.rewind();
+    ByteString encodedTags = ByteString.copyFrom(buffer, length);
```

Comment:
Not as far as I've been able to find. If the cost is prohibitive, we may ave to look into developing our own subclass implementation but we'll need some evidence first.